ruma-common/src/push/condition.rs

```diff
@@ -42,13 +49,203 @@ pub enum PushCondition {
+                        None => return false,
+                    },
+                };
+
+                matches_pattern(value, pattern, key == "content.body")
+            }
+            Self::ContainsDisplayName => {
+                let value = match event.get("content.body") {
+                    Some(v) => v,
+                    None => return false,
+                };
+
+                matches_pattern(value, &context.user_display_name, true)
+            }
+            Self::RoomMemberCount { is } => is.contains(&context.member_count),
+            Self::SenderNotificationPermission { key } => {
+                let sender_id = match event.get("sender") {
+                    Some(v) => match UserId::try_from(v.as_str()) {
+                        Ok(u) => u,
+                        Err(_) => return false,
+                    },
+                    None => return false,
+                };
+
+                let sender_level = match context.users_power_levels.get(&sender_id) {
+                    Some(i) => i,
+                    None => &context.default_power_level,
+                };
+
+                sender_level >= context.notification_power_levels.get(key)
+            }
+        }
+    }
+}
+
+/// The context of the room associated to an event to be able to test all push conditions
+#[derive(Clone, Debug)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct PushConditionRoomCtx {
+    /// The roomId of the room
+    pub room_id: String,
+
+    /// The number of members in the room
+    pub member_count: UInt,
+
+    /// The display name of the current user in the room
+    pub user_display_name: String,
+
+    /// The power levels of the users of the room
+    pub users_power_levels: BTreeMap<UserId, Int>,
+
+    /// The default power level of the users of the room
+    pub default_power_level: Int,
+
+    /// The notification power levels of the room
+    pub notification_power_levels: NotificationPowerLevels,
+}
+
+/// Matches the given value against a pattern.
+///
+/// The match is case insensitive.
+///
+/// If `match_words` is true, looks for pattern as a substring of `value`,
+/// and checks that it is a "word". If it is false, checks the pattern as
+/// a glob with wildcards `*` and `?`.
+fn matches_pattern(value: &str, pattern: &str, match_words: bool) -> bool {
+    if value.is_empty() || pattern.is_empty() {
+        return false;
+    }
+
+    let value = &value.to_uppercase();
+    let pattern = &pattern.to_uppercase();
+
+    if match_words {
+        matches_word(value, pattern)
+    } else {
+        WildMatch::new(pattern).matches(value)
+    }
+}
+
+fn matches_word(value: &str, pattern: &str) -> bool {
+    match value.find(pattern) {
+        Some(start) => {
+            let end = start + pattern.len();
+
+            // Look if the match has word boundaries
+            let word_boundary_start = {
+                if !char_at(value, start).is_alphanumeric() {
```

Comment:
All these uses of `is_alphanumeric` should be `is_word_char` instead (accounting for `_`s), no?

Reply:
Where does that method come from?

Reply:
We need to define it (`c.is_alphanumeric() || c == '_'). Could do an extension trait if you like the postfix notation. Same goes for `char_at` actually. `value.char_at(start)` seems easier to read, to me at least. But if you prefer free `fn`s, that works too.

Reply:
Ah right! I agree with you, trait extension would look nicer.

Reply:
Do I declare them in the same file or should I put them somewhere else?

Reply:
Also, is there a convention about naming the new trait?

Reply:
I'd just call it `StrExt` and put it in the same file. If it ends up being useful for other places too, then a more descriptive name would be in order but if it's just "a few str methods for use in push condition matching", `StrExt` inside `push/condition.rs` is perfectly fine.